File: 56-merge-intervals/merge-intervals.py
class Solution:
    def merge(self, intervals: list[list[int]]) -> list[list[int]]:
        # Sort by start time
        intervals.sort(key=lambda x: x[0])

        result = [intervals[0]]

        for start, end in intervals[1:]:
            last_start, last_end = result[-1]

            # If current interval overlaps with the last merged interval
            if start <= last_end:
                # Merge by extending the end
                result[-1][1] = max(last_end, end)
            else:
                # No overlap, add as new interval
                result.append([start, end])

        return result

File: 56-merge-intervals/merge-intervals.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        // Sort by start time
        sort(intervals.begin(), intervals.end());

        vector<vector<int>> result;
        result.push_back(intervals[0]);

        for (int i = 1; i < intervals.size(); i++) {
            int start = intervals[i][0];
            int end = intervals[i][1];
            int lastStart = result.back()[0];
            int lastEnd = result.back()[1];

            // If current interval overlaps with the last merged interval
            if (start <= lastEnd) {
                // Merge by extending the end
                result.back()[1] = max(lastEnd, end);
            } else {
                // No overlap, add as new interval
                result.push_back({start, end});
            }
        }

        return result;
    }
};